internal/static/cartogram2.js

```diff
@@ -630,9 +629,222 @@ class CartMap {
+     * @param {string} sysname The sysname of the map version
+     */
+    drawLegend(sysname){
+
+        // Get unit for the cartogram
+        const unit_carto = this.getLegendUnit(sysname);
+
+        // Obtain the original ratio of the values over total area.
+        const original_ratio = this.getTotalValuesForVersion(sysname)/this.getTotalAreaForVersion(sysname);
+
+        // square default is 30 by 30 px
+        const ratio = original_ratio*900;
+
+        // Calculate the number (10 to the power of something) that we need to round the ratio.
+        const round_ratio = Math.pow(10, (Math.round(ratio).toString().length-1));
+        // Round the ratio to a prettier number, such that they start with 1, 2, or 5
+        const r = ratio/round_ratio;
+        var final_ratio = 0;
+        if(Math.abs(r - 10) < Math.abs(r - 5) && Math.abs(r - 10) < Math.abs(r - 2)){
+            final_ratio = 10;
+        } else if (Math.abs(r - 5) < Math.abs(r - 2)){
+            final_ratio = 5;
+        } else {
+            final_ratio = 2;
+        }
+
+        // Calculate the scaled width and height of the square
+        const width = Math.sqrt(final_ratio*round_ratio*900/ratio);
+        let scale_word = (round_ratio > 999999) ? " million": round_ratio.toString().substr(1);
+        console.log("round_ratio" + round_ratio);
+        console.log("ratio" + ratio);
+        if(scale_word != " million" && scale_word.length >= 3){
+            const set_of_zeros = Math.floor(scale_word.length/3)
+            const remaining_zeros = scale_word.length%3
+            if(set_of_zeros == 1 && remaining_zeros == 0){
+                scale_word = "000".repeat(set_of_zeros)
+            } else{
+                scale_word = "0".repeat(remaining_zeros) + " 000".repeat(set_of_zeros);
+            }
+            console.log("carto scale word " + scale_word)
+        }
+        
+    
+        if(sysname != "1-conventional"){
+            document.getElementById('legend-square').setAttribute("width", width.toString() +"px");
+            document.getElementById('legend-square').setAttribute("height", width.toString() +"px");
+            document.getElementById("cartogram-legend").style.display = "inline";   
+
+            if(width > 35){
+                // adjust padding when the square is a bit too big
+                document.getElementById("legend-text").setAttribute("x", 50) 
+                document.getElementById("legend-text").innerHTML = "= " + final_ratio + scale_word + " " + unit_carto
+            } else{
+                document.getElementById("legend-text").innerHTML = "= " + final_ratio + scale_word + " " + unit_carto
+            }
+    
+        } else {
+            document.getElementById("cartogram-legend").style.display = "none";
+        }
     }
 
+    /**
+     * The following draws the legend for the conventional map
+     * @param {string} sysname The sysname of the map version
+     */
+    drawConvenLegend(sysname){
```

Comment:
If you make the suggested changes to the `drawLegend` function, then you do not need this function. You will be able to use the `drawLegend` function to draw the legend for any map version, conventional or cartogram.